src/Compilers/Core/Portable/Serialization/StreamObjectReader.cs

```diff
@@ -0,0 +1,1020 @@
+        {
+            return _reader.ReadSingle();
+        }
+
+        public override int ReadInt32()
+        {
+            return _reader.ReadInt32();
+        }
+
+        public override long ReadInt64()
+        {
+            return _reader.ReadInt64();
+        }
+
+        public override sbyte ReadSByte()
+        {
+            return _reader.ReadSByte();
+        }
+
+        public override short ReadInt16()
+        {
+            return _reader.ReadInt16();
+        }
+
+        public override uint ReadUInt32()
+        {
+            return _reader.ReadUInt32();
+        }
+
+        public override ulong ReadUInt64()
+        {
+            return _reader.ReadUInt64();
+        }
+
+        public override ushort ReadUInt16()
+        {
+            return _reader.ReadUInt16();
+        }
+
+        public override DateTime ReadDateTime()
+        {
+            return DateTime.FromBinary(_reader.ReadInt64());
+        }
+
+        public override string ReadString()
+        {
+            return ReadStringValue();
+        }
+
+        public override object ReadValue()
+        {
+            var v = ReadVariant();
+
+            // if we didn't get anything, it must have been an object or array header
+            if (v.Kind == VariantKind.None)
+            {
+                v = ConsumeAndConstruct();
+            }
+
+            return v.ToBoxedObject();
+        }
+
+        private Variant ConsumeAndConstruct()
+        {
+            Debug.Assert(_consumerStack.Count > 0);
+
+            // keep reading until we've got all the elements to construct the object or array
+            while (_consumerStack.Count > 0)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                var consumer = _consumerStack.Peek();
+                if (consumer.ElementCount > 0 && _valueStack.Count < consumer.StackStart + consumer.ElementCount)
+                {
+                    var element = ReadVariant();
+                    if (element.Kind != VariantKind.None)
+                    {
+                        _valueStack.Push(element);
+                    }
+                }
+                else
+                {
+                    consumer = _consumerStack.Pop();
+
+                    var constructed = (consumer.Reader != null)
+                        ? ConstructObject(consumer.Type, consumer.ElementCount, consumer.Reader, consumer.Id)
+                        : ConstructArray(consumer.Type, consumer.ElementCount);
+
+                    _valueStack.Push(constructed);
+                }
+            }
+
+            Debug.Assert(_valueStack.Count == 1);
+            return _valueStack.Pop();
+        }
+
+        private struct Consumer
+        {
+            public readonly Type Type;
+            public readonly int ElementCount;
+            public readonly int StackStart;
+            public readonly Func<ObjectReader, object> Reader;
```

Comment:
This type is shared for both object and array deserialization.  Only object deserialiation has a reader.